matchbox_signaling/src/signaling_server/builder.rs

```diff
@@ -0,0 +1,202 @@
+where
+    Topology: SignalingTopology<Cb, S>,
+    Cb: SignalingCallbacks,
+    S: SignalingState,
+{
+    /// Creates a new builder for a [`SignalingServer`].
+    pub fn new(socket_addr: impl Into<SocketAddr>, topology: Topology, state: S) -> Self {
+        Self {
+            socket_addr: socket_addr.into(),
+            router: Router::new()
+                .route("/", get(ws_handler::<Cb, S>))
+                .route("/:path", get(ws_handler::<Cb, S>))
+                .with_state(state),
+            callbacks: Cb::default(),
+            topology,
+            state: PhantomData,
+        }
+    }
+
+    /// Modify the router with a mutable closure. This is where one may apply middleware or other
+    /// layers to the Router.
+    pub fn middleware(mut self, mut alter: impl FnMut(&mut Router)) -> Self {
+        alter(&mut self.router);
+        self
+    }
+
+    /// Change the topology.
+    pub fn topology(mut self, topology: Topology) -> Self {
+        self.topology = topology;
+        self
+    }
+
+    /// Apply permissive CORS middleware for debug purposes.
+    pub fn cors(mut self) -> Self {
+        self.router = self.router.layer(
+            CorsLayer::new()
+                .allow_origin(Any)
+                .allow_methods(Any)
+                .allow_headers(Any),
+        );
+        self
+    }
+
+    /// Apply a default tracing middleware layer for debug purposes.
+    pub fn trace(mut self) -> Self {
+        self.router = self.router.layer(
+            // Middleware for logging from tower-http
+            TraceLayer::new_for_http().on_response(
+                DefaultOnResponse::new()
+                    .level(Level::INFO)
+                    .latency_unit(LatencyUnit::Micros),
+            ),
+        );
+        self
+    }
+
+    /// Create a [`SignalingServer`].
+    ///
+    /// # Panics
+    /// This method will panic if the socket address requested cannot be bound.
+    pub fn build(mut self) -> SignalingServer {
+        // Insert topology
+        let state_machine: SignalingStateMachine<Cb, S> =
+            SignalingStateMachine::from_topology(self.topology);
+        self.router = self
+            .router
+            .layer(Extension(state_machine))
+            .layer(Extension(self.callbacks))
+            .layer(Extension(self.state));
+        let server = axum::Server::bind(&self.socket_addr).serve(
+            self.router
+                .into_make_service_with_connect_info::<SocketAddr>(),
+        );
+        let socket_addr = server.local_addr();
+        SignalingServer {
+            server,
+            socket_addr,
+        }
+    }
+}
+
+impl SignalingServerBuilder<FullMesh, FullMeshCallbacks, FullMeshState> {
+    /// Set a callback triggered on signals.
+    pub fn on_signal<F>(mut self, callback: F) -> Self
+    where
+        F: Fn(JsonPeerRequest) + 'static,
+    {
+        self.callbacks.on_signal = Callback::from(callback);
+        self
+    }
```

Comment:
What would be the use case of accessing the signal payload?

Reply:
Not sure why someone would want this. Maybe they just want to log all the signals.

Reply:
I think this is perhaps already covered by `trace` logging? Maybe remove it until somebody has a need for it? Think this would get rid of the need to clone the signal as well.